README.md
# slackin

A little server that enables public access to a Slack server. Like Freenode, but on Slack.

It provides

- A landing page you can point users to fill in their
  emails and receive an invite (`https://slack.yourdomain.com`)
- An `<iframe>` badge to embed on any website
  that shows connected users in *realtime* with socket.io.
- A SVG badge that works well from static mediums
  (like GitHub README pages)

Read more about the [motivations and history](http://rauchg.com/slackin) behind Slackin.

## How to use

### Server

#### Heroku

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/rauchg/slackin/tree/master)

#### Azure

[![Deploy to Azure](http://azuredeploy.net/deploybutton.svg)](https://azuredeploy.net/)

#### OpenShift

[Follow these instructions.](https://github.com/rauchg/slackin/wiki/OpenShift)

#### Cloud Foundry

##### IBM Bluemix
[![Deploy to Bluemix](https://bluemix.net/deploy/button.png)](https://bluemix.net/deploy?repository=https://github.com/rauchg/slackin)

For other CloudFoundry providers, [follow these instructions.](https://github.com/pivotal-cf/slackin/wiki/Cloud-Foundry)

#### NPM

Install it and launch it on your server:

```bash
$ npm install -g slackin
$ slackin "your-team-id" "your-slack-token"
```

Your team id is what you use to access your login page on Slack (eg: https://**{this}**.slack.com).

You can find or generate your API test token at [api.slack.com/web](https://api.slack.com/web) – note that the user you use to generate the token must be an admin. You need to create a dedicated `@slackin-inviter` user (or similar), mark that user an admin, and use a test token from that dedicated admin user.  Note that test tokens have actual permissions so you do not need to create an OAuth 2 app. Also check out the Slack docs on [generating a test token](https://get.slack.help/hc/en-us/articles/215770388-Creating-and-regenerating-API-tokens).

The available options are:

```
Usage: slackin [options] <team-id> <api-token>

Options:

  -?, --help                 output usage information
  -v, --version              output the version number
  -p, --port <port>          Port to listen on [$PORT or 3000]
  -h, --hostname <hostname>  Hostname to listen on [$HOSTNAME or 0.0.0.0]
  -c, --channels [<chan>]    One or more comma-separated channel names to allow single-channel guests [$SLACK_CHANNELS]
  -c, --channel <chan>       Single channel guest invite (deprecated) [$SLACK_CHANNEL]
  -i, --interval <int>       How frequently (ms) to poll Slack [$SLACK_INTERVAL or 5000]
  -P, --path                 Path to serve slackin under
  -s, --silent               Do not print out warns or errors
  -c, --css <file>           Full URL to a custom CSS file to use on the main page
```

**Important: if you use Slackin in single-channel mode, you'll only be
able to invite as many external accounts as paying members you have
times 5. If you are not getting invite emails, this might be the reason.
Workaround: sign up for a free org, and set up Slackin to point to it
(all channels will be visible).**

### Realtime Badge

[![](https://cldup.com/IaiPnDEAA6.gif)](http://slack.socket.io)

```html
<script async defer src="https://slack.yourdomain.com/slackin.js"></script>
```

or for the large version, append `?large`:

```html
<script async defer src="https://slack.yourdomain.com/slackin.js?large"></script>
```

### SVG

[![](https://cldup.com/jWUT4QFLnq.png)](http://slack.socket.io)

```html
<img src="https://slack.yourdomain.com/badge.svg">
```

Done in Markdown this looks like:

    [![Slack Status](https://slack.yourdomain.com/badge.svg)](https://yourdomain.com)

### Landing page

[![](https://cldup.com/WIbawiqp0Q.png)](http://slack.socket.io)

Point to `https://slack.yourdomain.com`.

**Note:** the image for the logo of the landing page
is retrieved from the Slack API. If your organization
doesn't have one configured, it won't be shown.

## API

Requiring `slackin` as a module will return
a `Function` that creates a `HTTP.Server` instance
that you can manipulate.

```js
require('slackin').default({
  token: 'yourtoken', // required
  interval: 1000,
  org: 'your-slack-subdomain', // required
  path: '/some/path/you/host/slackin/under/', // defaults to '/'
  channels: 'channel,channel,...' // for single channel mode
  silent: false // suppresses warnings
}).listen(3000);
```

This will show response times from Slack and how many
online users you have on the console.

By default logging is enabled.

The returned `http.Server` has an `app` property that is
the `express` application that you can define or override
routes on.

### JSON

All the metadata for your organization can be fetched
via a JSON HTTP request to `/data`.

If you wish to turn on CORS, pass `-x` or `--cors` to `slackin`.

## Developing

Slackin's server side code is written in ES6. It uses babel to transpile the
ES6 code to a format node understands. After cloning Slackin, you should
install the prerequisite node libraries with npm:

```bash
$ npm install
```

After the libraries install, the postinstall script will run `gulp` to invoke
babel on the source. It is important to run `gulp` manually after updating any
files in lib/ to update the versions in node/.

## Credits

- The SVG badge generation was taken from the
excellent [shields](https://github.com/badges/shields) project.
- The button CSS is based on
[github-buttons](https://github.com/mdo/github-buttons).

## License

MIT


HISTORY.md

0.13.0 / 2016-07-20
===================

  * index: add `/data` endpoint to get data through a JSON API

0.12.0 / 2016-07-20
===================

  * add options cors support
  * slack: improve error handling

0.11.3 / 2016-07-19
===================

  * package: don't forget about assets

0.11.2 / 2016-07-19
===================

  * package: fix broken `main`

0.11.1 / 2016-07-19
===================

  * package: fix usecase of `require('slackin')`

0.11.0 / 2016-07-19
===================

  * index: expose express app as `app` on the returned value of `slackin`

0.10.3 / 2016-07-19
===================

  * package: only publish `bin` and `dist`

0.10.2 / 2016-07-19
===================

  * package: remove postinstall

0.10.1 / 2016-07-19
===================

  * moving babel and gulp into dev deps [@rauchg]
  * package: remove `start` as it requires ENV stuff [@rauchg]

0.10.0 / 2016-07-19
===================

  * package: bump `socket.io` for security fixes

0.9.0 / 2016-07-19
==================

  * Important bug fixes (#208) [@leo]
  * Switch to args, lint code and clean up (#200) [@leo]
  * Update docs with more clear token generation info
  * Change heroku deploy to master
  * Update Dockerfile to match Procfile
  * issue #169, add turnary assignment for client sdk channels
  * Change case on the "already invited" message
  * Update Readme.md
  * Fix TypeError when submitting w/o SLACK_CHANNELS config

0.8.3 / 2016-03-08
==================

  * fix postMessage({}, '*') causes slackin to throw [@laughinghan]
  * add IBM Bluemix Deploy badge [@kevinSuttle]
  * add `?large` to `/iframe` and `/iframe/dialog` [@laughinghan]
  * convert dialog dimensions to em/rem for adjustability #prepwork [@laughinghan]
  * fix redirecting already-registered users from iframe dialog [@laughinghan]
  * add "or sign in" link to iframe too [@laughinghan]
  * render a hidden form field for a single channel [@nickstenning]
  * add .travis.yml [@jszwedko]
  * fix test for successful invite [@jszwedko]
  * update `babel-register` module name in mocha opts [@jszwedko]
  * update readme for babel 6 change [@danreeves]
  * shorten already-invited message [@MaxWofford]
  * add success message for already signed up users [@MaxWofford]
  * fix max listeners warnings [Hardeep Shoker]

0.8.2 / 2016-01-13
==================

  * use better supported anonymous function syntax [WouterSioen]
  * bump socket.io

0.8.1 / 2016-01-06
==================

  * index: fix `import` order issue

0.8.0 / 2016-01-06
==================

  * bump socket.io to 1.4.0 for sec advisory
  * add deploy to Azure Button
  * build: Update builds to use gulp and babel6 (Fixes #122)
  * adds instructions for deploying to Cloud Foundry
  * redirect signed up users to Slack
  * fix --path option parsing

0.7.3 / 2015-11-29
==================

  * package: bump `socket.io` to fix build issue

0.7.2 / 2015-11-29
==================

  * fix rendering of badge in non-chrome

0.7.1 / 2015-11-19
==================

  * fix inline badge styles

0.7.0 / 2015-11-19
==================

  * badge: make badge work for users with lots of users
  * splash: fix hover color – thank goodness
  * relative path hosting support
  * fix tests
  * add support for codes of conduct

0.6.0 / 2015-10-05
==================

  * use faster API methods: users.list to get presence
  * fix polling interval increase
  * allow hosting slackin under a path (eg: /slack)
  * fix for deployment instructions
  * add already registered link
  * use SLACK_CHANNELS instead of SLACK_CHANNEL on Heroku
  * standardize use of example host
  * allow hosting at a subfolder

0.5.1 / 2015-06-23
==================

  * added safety check for channels.
  * fix counting deleted users

0.5.0 / 2015-05-07
==================

  * allow for multiple guest channels [afeld]
  * add important notes to README
  * better error handling and tests [afeld]
  * add Slack favicon and better title [frabrunelle]
  * error when the token account is not an admin [martypenner]
  * added Dockerfile [darron]

0.4.0 / 2015-04-10
==================

 * use SVG Heroku button
 * fixed svg dimensions issues
 * exclude bots (and slackbot) from users count
 * improve mobile zoom
 * use `<input type="email">`
 * fix for single-channel guests invitation parameter name

0.3.1 / 2015-03-16
==================

 * fix meta tags for mobile support

0.3.0 / 2015-03-06
==================

 * update `babel`, fix all scenarios

0.2.9 / 2015-03-03
==================

 * package: bump `socket.io`

0.2.8 / 2015-02-03
==================

 * bump socket.io

0.2.7 / 2015-02-02
==================

 * support for custom css

0.2.6 / 2015-02-01
==================

 * fix issue with 6to5 and `-g install`

0.2.5 / 2015-02-01
==================

 * fix single channel invites to be ultra restricted

0.2.4 / 2015-02-01
==================

 * bump 6to5

0.2.3 / 2015-02-01
==================

 * fixed heroku button

0.2.2 / 2015-01-30
==================

 * fixed ability to `require('slackin')`

0.2.1 / 2015-01-30
==================

 * improved startup speed by leveraging pre-compilation

0.2.0 / 2015-01-30
==================

 * improved CLI interface [buildkite]
 * support for ENV vars [buildkite]
 * deploy to heroku button [buildkite]

0.1.2 / 2015-01-28
==================

 * fix bin name [brianloweswords]

0.1.1 / 2015-01-26
==================

 * fix typo

0.1.0 / 2015-01-26
==================

 * initial release


lib/index.js
// es6 runtime requirements
import 'babel-polyfill'

// their code
import express from 'express'
import sockets from 'socket.io'
import { json } from 'body-parser'
import { Server as http } from 'http'
import remail from 'email-regex'
import dom from 'vd'
import cors from 'cors'

// our code
import Slack from './slack'
import invite from './slack-invite'
import badge from './badge'
import splash from './splash'
import iframe from './iframe'
import log from './log'

export default function slackin ({
  token,
  interval = 5000, // jshint ignore:line
  org,
  css,
  coc,
  cors: useCors = false,
  path='/',
  channels,
  silent = false // jshint ignore:line,
}){
  // must haves
  if (!token) throw new Error('Must provide a `token`.')
  if (!org) throw new Error('Must provide an `org`.')

  if (channels) {
    // convert to an array
    channels = channels.split(',').map((channel) => {
      // sanitize channel name
      if ('#' === channel[0]) return channel.substr(1)
      return channel
    })
  }

  // setup app
  let app = express()
  let srv = http(app)
  srv.app = app

  let assets = __dirname + '/assets'

  // fetch data
  let slack = new Slack({ token, interval, org })

  slack.setMaxListeners(Infinity)

  // capture stats
  log(slack, silent)

  // middleware for waiting for slack
  app.use((req, res, next) => {
    if (slack.ready) return next()
    slack.once('ready', next)
  })

  if (useCors) {
    app.options('*', cors())
    app.use(cors())
  }

  // splash page
  app.get('/', (req, res) => {
    let { name, logo } = slack.org
    let { active, total } = slack.users
    if (!name) return res.send(404)
    let page = dom('html',
      dom('head',
        dom('title',
          'Join ', name, ' on Slack!'
        ),
        dom('meta name=viewport content="width=device-width,initial-scale=1.0,minimum-scale=1.0,user-scalable=no"'),
        dom('link rel="shortcut icon" href=https://slack.global.ssl.fastly.net/272a/img/icons/favicon-32.png'),
        css && dom('link rel=stylesheet', { href: css })
      ),
      splash({ coc, path, css, name, org, logo, channels, active, total })
    )
    res.type('html')
    res.send(page.toHTML())
  })

  app.get('/data', (req, res) => {
    let { name, logo } = slack.org
    let { active, total } = slack.users
    res.send({
      name,
      org,
      coc,
      logo,
      channels,
      active,
      total
    })
  })

  // static files
  app.use('/assets', express.static(assets))

  // invite endpoint
  app.post('/invite', json(), (req, res, next) => {
    let chanId
    if (channels) {
      let channel = req.body.channel
      if (!channels.includes(channel)) {
        return res
        .status(400)
        .json({ msg: 'Not a permitted channel' })
      }
      chanId = slack.getChannelId(channel)
      if (!chanId) {
        return res
        .status(400)
        .json({ msg: `Channel not found "${channel}"` })
      }
    }

    let email = req.body.email

    if (!email) {
      return res
      .status(400)
      .json({ msg: 'No email provided' })
    }

    if (!remail().test(email)) {
      return res
      .status(400)
      .json({ msg: 'Invalid email' })
    }

    if (coc && '1' != req.body.coc) {
      return res
      .status(400)
      .json({ msg: 'Agreement to CoC is mandatory' })
    }

    invite({ token, org, email, channel: chanId }, err => {
      if (err) {
        if (err.message === `Sending you to Slack...`) {
          return res
          .status(303)
          .json({ msg: err.message, redirectUrl: `https://${org}.slack.com` })
        }

        return res
        .status(400)
        .json({ msg: err.message })
      }

      res
      .status(200)
      .json({ msg: 'WOOT. Check your email!' })
    })
  })

  // iframe
  app.get('/iframe', (req, res) => {
    let large = 'large' in req.query
    let { active, total } = slack.users
    res.type('html')
    res.send(iframe({ path, active, total, large }).toHTML())
  })

  app.get('/iframe/dialog', (req, res) => {
    let large = 'large' in req.query
    let { name } = slack.org
    let { active, total } = slack.users
    if (!name) return res.send(404)
    let dom = splash({ coc, path, name, org, channels, active, total, large, iframe: true })
    res.type('html')
    res.send(dom.toHTML())
  })

  // badge js
  app.use('/slackin.js', express.static(assets + '/badge.js'))

  // badge rendering
  app.get('/badge.svg', (req, res) => {
    res.type('svg')
    res.set('Cache-Control', 'max-age=0, no-cache')
    res.set('Pragma', 'no-cache')
    res.send(badge(slack.users).toHTML())
  })

  // realtime
  sockets(srv).on('connection', socket => {
    socket.emit('data', slack.users)
    let change = (key, val) => socket.emit(key, val)
    slack.on('change', change)
    socket.on('disconnect', () => {
      slack.removeListener('change', change)
    })
  })

  return srv
}
